src/graphics/pyro/pyro.cpp

```diff
@@ -0,0 +1,480 @@
+}
+
+void CPyro::UpdateEffect()
+{
+}
+
+Error CPyro::IsEnded()
+{
+    // End of the pyrotechnic effect?
+    if ( m_progress < 1.0f )  return ERR_CONTINUE;
+
+    return ERR_STOP;
+}
+
+void CPyro::AfterEnd()
+{
+}
+
+void CPyro::CutObjectLink(CObject* obj)
+{
+    if (m_object == obj)
+        m_object = nullptr;
+}
+
+void CPyro::CreateLight(Math::Vector pos, float height)
+{
+    if (!m_engine->GetLightMode()) return;
+
+    m_lightHeight = height;
+
+    Gfx::Light light;
+    light.type        = LIGHT_SPOT;
+    light.ambient     = Gfx::Color(0.0f, 0.0f, 0.0f);
+    light.position    = Math::Vector(pos.x, pos.y+height, pos.z);
+    light.direction   = Math::Vector(0.0f, -1.0f, 0.0f);  // against the bottom
+    light.spotIntensity = 1.0f;
+    light.attenuation0 = 1.0f;
+    light.attenuation1 = 0.0f;
+    light.attenuation2 = 0.0f;
+    light.spotAngle = Math::PI/4.0f;
+
+    m_lightRank = m_lightMan->CreateLight();
+
+    m_lightMan->SetLight(m_lightRank, light);
+    m_lightMan->SetLightIntensity(m_lightRank, 0.0f);
+
+    // Only illuminates the objects on the ground.
+    m_lightMan->SetLightIncludeType(m_lightRank, ENG_OBJTYPE_TERRAIN);
+}
+
+void CPyro::DeleteObject(bool primary, bool secondary)
+{
+    if (m_object == nullptr) return;
+
+    ObjectType type = m_object->GetType();
+    if ( secondary              &&
+         type != OBJECT_FACTORY &&
+         type != OBJECT_NUCLEAR &&
+         type != OBJECT_ENERGY )
+    {
+        if (m_object->Implements(ObjectInterfaceType::Powered))
+        {
+            CPoweredObject& poweredObject = dynamic_cast<CPoweredObject&>(*m_object);
+            CObject* sub = poweredObject.GetPower();
+            if (sub != nullptr)
+            {
+                CObjectManager::GetInstancePointer()->DeleteObject(sub);
+                poweredObject.SetPower(nullptr);
+            }
+        }
+
+        if (m_object->Implements(ObjectInterfaceType::Carrier))
+        {
+            CCarrierObject& carrierObject = dynamic_cast<CCarrierObject&>(*m_object);
+            CObject* sub = carrierObject.GetCargo();
+            if (sub != nullptr)
+            {
+                CObjectManager::GetInstancePointer()->DeleteObject(sub);
+                carrierObject.SetCargo(nullptr);
+            }
+        }
+    }
+
+    if (primary)
+    {
+        if (m_object->Implements(ObjectInterfaceType::Transportable))
+        {
+            // TODO: this should be handled in the object's destructor
+            CObject* transporter = dynamic_cast<CTransportableObject&>(*m_object).GetTransporter();
+            if (transporter != nullptr)
+            {
+                if (m_object->Implements(ObjectInterfaceType::Powered))
+                {
+                    CObject* sub = dynamic_cast<CPoweredObject&>(*m_object).GetPower();
+                    if (sub != nullptr)  // is there a battery?
+                        sub->SetScaleY(1.0f - m_progress);  // complete flattening
+                }
```

Comment:
Please explain this diff:
```diff
- if (transporter->Implements(ObjectInterfaceType::Powered))
- {
-    CPoweredObject* powered = dynamic_cast<CPoweredObject*>(transporter);
-    if (powered->GetPower() == m_object)
-        powered->SetPower(nullptr);
-}
+ if (m_object->Implements(ObjectInterfaceType::Powered))
+ {
+     CObject* sub = dynamic_cast<CPoweredObject&>(*m_object).GetPower();
+     if (sub != nullptr)  // is there a battery?
+         sub->SetScaleY(1.0f - m_progress);  // complete flattening
+ }
```

The code you deleted roughly translates to "If m_object has a parent object and if m_object is installed in the parent's power cell slot, set the parent's m_power pointer to null to prevent dangling pointers". The code you replaced it with roughly translates to "If m_object has a parent object and if **m_object** has a power cell, flatten the power cell". Either there is something I deeply misunderstood about how this works or this is complete nonsense.